### PR TITLE
Adds an Ozone Implementation archetype

### PR DESCRIPTION
**Pull request by ibacher**
This adds an Maven archetype that creates a skeleton Ozone Implementation project

### CODE REVIEW

**Comment by ibacher** [comment]
I don't know why the validate step is failing. @corneliouzbett any ideas?

**Comment by corneliouzbett** [comment]
> I don't know why the validate step is failing. @corneliouzbett any ideas?

Secrets are not passed to the runner when a workflow is triggered from a forked repository. So both username and password will be empty in the settings.xml. This is a security measure to prevent unauthorized access to sensitive data.

I believe, we don't need secrets for the validation step for this repo but we need for private repos.

**Comment by rbuisson** [comment]
@corneliouzbett @ibacher , the [PR](https://github.com/mekomsolutions/shared-github-workflow/pull/26) to fix the workflow is merged.

**Review by rbuisson — COMMENTED** [review]
- Would it be possible to have the interactive mode to default the version to `1.0.0-SNAPSHOT` instead of `1.0-SNAPSHOT`?
```Define value for property 'version' 1.0-SNAPSHOT:```

- The default location for config files for OpenMRS for instance should be `configs/openmrs/(frontend_config|initializer_config)/`.

- It sounds like this is a distro-only archetype. Shouldn't it be a complete Ozone archetype which sets up the `distro` sub module but will also bring the `run/` resources upon build?

**Comment by ibacher** [comment]
For the tests here to work, I need the Maven job to run the `install` step and not _just_ the verify step. Is there a way to do that?

**Comment by corneliouzbett** [comment]
> For the tests here to work, I need the Maven job to run the install step and not just the verify step. Is there a way to do that?

Yes, by modifying [this](https://github.com/mekomsolutions/shared-github-workflow/blob/main/.github/workflows/maven-build-test.yml#L78) to run `install` or take input variable `maven-phase` which runs `verify` by default.

**Review by rbuisson — CHANGES_REQUESTED** [review]

**Comment by rbuisson** [comment]
> The idea of the archetype is to make it easier to setup distro-specific projects, since that's what there will be multiple versions of, so this is designed to produce something like ozone-distro-cambodia or ozone-kenyahmis. Do we have a use-case for multiple things with a distro submodule that pull in the run resources on build?

@ibacher , so I just want to make sure that by default the "distribution" that will use this artefact will inherit/use/depend on the whole `ozone` project, not only the `ozone-distro` part, as we need the projects to be able to run the suite.

**Comment by rbuisson** [comment]
>We can probably abuse the fact that they're all submodules of a single project to do something.

> This is modeled pretty concretely on what we're already doing with distributions (go-to-scripts-dir.sh needs to be filtered and then set-ozone-dir.sh relies on go-to-scripts-dir.sh being in the same folder), so it's not added maintenance over-and-above what already happens with ozone-cambodia or ozone-kenyahmis.

Well, it does add a little bit of maintenance in the sense that it wasn't ours to do before. Now the file is hosted in the repo, that's upon us to maintain the 2 versions.
I'm ok with having this duplication right now. And see to improve it later. Anyway some of those scripts are already duplicated in Ozone Docker Compose too. A little piece of work there.

**Comment by ibacher** [comment]
@rbuisson So the projects created by this archteyp use the ozone-distro-parent as a parent. ozone-distro-parent adds the full Ozone distribution as a dependency and the various default steps to work with it.

File: README.md [hunk 1]
```diff
@@ -63,3 +63,22 @@ For example:
...
+Once that's done, you can use the Maven Archetype by running:
+
+```bash
+mvn archetype:generate -DarchetypeGroupId=com.ozonehis -DarchetypeArtifactId=ozone-distro-archetype
+```
+
+This will prompt you for details about your project and create a standard Maven repository in the local directory.
+
+**Note** You do **not** want your distribution to be part of the Ozone source code, so you should create it in a directory
+that does not contain this project.
```
[the remark this answers is not in the document]
Comment (ibacher) [reply README.md]: Yeah... IDE autoformatting there.

File: ozone-distro-archetype/src/main/resources/archetype-resources/readme/impl-guide.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+#### Option 1. Turn down the whole project and start afresh
+```bash
+source target/go-to-scripts-dir.sh
+./destroy-demo.sh
+```
+
+Re-build:
+```bash
+./scripts/mvnw clean package
```
Comment (rbuisson) [inline ozone-distro-archetype/src/main/resources/archetype-resources/readme/impl-guide.md:32]: This seems to assume the Maven Wrapper is in **scripts/** though, isn't it?

**Comment by rbuisson** [comment]
> Yeah, that's trivial to do, although I think putting the wrapper in a sub-directory is a little odd.

I wanted to keep the root directory as clean as possible. But if that's an unusual thing to do and Maven Wrapper is 99% of the time located at the root folder, let's do have it there. We can modify Ozone itself accordingly in a follow up commit.

File: ozone-distro-archetype/src/main/resources/archetype-resources/readme/impl-guide.md (hunk 2 above)
Comment (ibacher) [reply ozone-distro-archetype/src/main/resources/archetype-resources/readme/impl-guide.md:32]: I moved it and updated the text here.

File: README.md [hunk 3]
```diff
@@ -63,3 +63,22 @@ For example:
...
 
 💡 **Did you know?** Ozone Pro comes with single sign-on (SSO) and all its integration layer is secured with OAuth2.
+
+## Create Your Own Distribution
+
+This Ozone project provides tooling to help create your own distribution. To get started, you first need to install the
+archetype by running:
+
+```bash
+./scripts/mvn -pl ozone-distro-archetype clean install
```
Comment (rbuisson) [inline README.md:73]: I guess `./scripts/mvnw`, right?

File: maven-archetype/src/main/resources/archetype-resources/readme/impl-guide.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#  Ozone ${distributionName} - Implementer Guide
+
+This distribution can be run using the Ozone Docker project, which is the default configuration for this. The quick start command below is for demonstration and trial purposes and would not be suitable for a stable environment.
+
+## Quick Start
+
+Build
+```bash
+./mvnw clean package
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/readme/impl-guide.md:9]: `./mvnw clean package`
Not sure anymore, do we install Maven Wrapper in root or in `scripts/`

File: maven-archetype/src/main/resources/archetype-resources/pom.xml [hunk 5]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>com.ozonehis</groupId>
+    <artifactId>ozone-distro-parent</artifactId>
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/pom.xml:7]: I'm guessing it should be now `maven-parent`, right?

File: maven-archetype/pom.xml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>com.ozonehis</groupId>
+    <artifactId>ozone-parent</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+    <relativePath>../ozone-parent</relativePath>
```
Comment (rbuisson) [inline maven-archetype/pom.xml:8]: `../maven-parent`

File: maven-archetype/pom.xml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>com.ozonehis</groupId>
+    <artifactId>ozone-parent</artifactId>
```
Comment (rbuisson) [inline maven-archetype/pom.xml:6]: `maven-parent`

File: maven-archetype/src/main/resources/archetype-resources/pom.xml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    <version>1.0.0-SNAPSHOT</version>
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/pom.xml:8]: Any way to manage this version of the parent to the current version in `../maven-parent/` ?

File: README.md [hunk 9]
```diff
@@ -63,3 +63,22 @@ For example:
 Ozone FOSS requires you to log into each component separately.
 
 💡 **Did you know?** Ozone Pro comes with single sign-on (SSO) and all its integration layer is secured with OAuth2.
+
+## Create Your Own Distribution
```
Comment (rbuisson) [inline README.md:67]: Would you be able to add this section as a PR in the https://github.com/ozone-his/ozone-docs repo instead?

File: maven-archetype/src/main/resources/archetype-resources/README.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+# Ozone ${distributionName}
+
+**Ozone ${distributionName}** is a distribution of [Ozone HIS](https://www.ozone-his.com).
+
+## Implementer Guide
+A technical guide to help implementers building and running the project can be found [here](readme/impl-guide.md).
+
+## Release Notes
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/README.md:8]: Could you add this under release notes:

```
### ${version} (in progress)

- 
```

File: maven-archetype/src/main/resources/archetype-resources/readme/impl-guide.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
+#  Ozone ${distributionName} - Implementer Guide
+
+This distribution can be run using the Ozone Docker project, which is the default configuration for this. The quick start command below is for demonstration and trial purposes and would not be suitable for a stable environment.
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/readme/impl-guide.md:3]: ```using the [Ozone Docker Compose](https://github.com/ozone-his/ozone-docker-compose) project```

**Review by rbuisson — CHANGES_REQUESTED** [review]

File: maven-archetype/src/main/resources/archetype-resources/readme/impl-guide.md (hunk 4 above)
Comment (ibacher) [reply maven-archetype/src/main/resources/archetype-resources/readme/impl-guide.md:9]: Scripts now. Fixed.

File: maven-archetype/src/main/resources/archetype-resources/pom.xml (hunk 8 above)
Comment (ibacher) [reply maven-archetype/src/main/resources/archetype-resources/pom.xml:8]: Not as directly as I'd like. What I can do (pretty easily) is add a step to the post-generate script that will set this version to match the version of the maven-archetype itself...

File: maven-archetype/src/main/resources/archetype-resources/.gitpod.yml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+tasks:
+  - name: Run Ozone ${distributionName}
+    before: sudo apt-get update && sudo apt-get install -y gettext-base && sudo rm -rf /var/lib/apt/lists/*
+    init: mvnw clean package
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/.gitpod.yml:4]: Prob needs to be:
```
scripts/mvnw clean package
```

File: maven-archetype/src/main/resources/META-INF/archetype-post-generate.groovy [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    def process = command.execute(null, projectDirectory.toFile());
+    process.consumeProcessOutput(System.out, System.err)
+    process.waitFor()
+    if (process.exitValue() != 0) {
+        throw new RuntimeException("'$command' exited with code ${process.exitValue()}")
+    }
+}
+
+// add the Maven wrapper
+run("mvn wrapper:wrapper")
```
Comment (rbuisson) [inline maven-archetype/src/main/resources/META-INF/archetype-post-generate.groovy:16]: What version of Maven is running here? Does this assume Maven to be installed on the machine?

The fact that the project build log looses the coloring when running this command made me wonder...
Screenshot 2024-02-12 at 07 59 43

Would it be possible to use the Wrapper provided by the project to run the commands instead?

**Comment by rbuisson** [comment]
For some reason the final package ends up missing the **readme/impl-guide.md** file.
```
myOzone/
├── README.md
├── config
│   └── openmrs
│       ├── frontend_config
│       └── initializer_config
├── pom.xml
└── scripts
    ├── go-to-scripts-dir.sh
    ├── mvnw
    ├── mvnw.cmd
    ├── set-ozone-dir.sh
    └── start-ozone.sh

File: maven-archetype/src/main/resources/META-INF/archetype-post-generate.groovy (hunk 13 above)
Comment (ibacher) [reply maven-archetype/src/main/resources/META-INF/archetype-post-generate.groovy]: So, it turns out to be pretty hard to do this. If I _just_ do this with the wrapper from this repo, then the script won't work when used outside of this repo (it'd also get pretty gnarly to make integration test pass). I was hoping to be able to get access to the Maven session, so I could just execute things in the current Maven context, but that isn't available inside the script.

I've resorted to the work-around where I've basically re-implemented the `mvn wrapper:wrapper` part in Groovy, using Groovy built-ins. Then, once the wrapper is installed, we use that for any additional Maven commands. The script is a little verbose, but it should basically work.

**Comment by ibacher** [comment]
Re: **readme/impl-guide.md**: My fault. The archetype only picks up on files described in the `src/main/resources/META-INF/maven/archetype-metadata.xml` descriptor. I mistakenly left that out.

**Comment by ibacher** [comment]
Also, the scripts have been removed from this PR, but they're added back with #42.

**Comment by rbuisson** [comment]
Looks pretty good! Thanks.

File: maven-archetype/src/main/resources/archetype-resources/.gitignore
Comment (rbuisson) [inline maven-archetype/src/main/resources/archetype-resources/.gitignore:1]: This file does not seem to be included in the final package. Is this intended?

Comment (ibacher) [reply maven-archetype/src/main/resources/archetype-resources/.gitignore:1]: No... I guess it's [this issue](https://issues.apache.org/jira/projects/ARCHETYPE/issues/ARCHETYPE-505). Adding a work-around.

**Comment by rbuisson** [comment]
@ibacher  do you plan to send a PR on [Ozone Docs](https://github.com/ozone-his/ozone-docs) too?

**Review by rbuisson — APPROVED** [review]

**Comment by ibacher** [comment]
@rbuisson PR for docs is [here](https://github.com/ozone-his/ozone-docs/pull/1).

**Pull request merged**